crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +129,113 @@ impl<S: States> NextState<S> {
     }
 }
 
+/// The previous state of [`State<S>`].
+///
+/// This is used internally by Bevy and generally not meant to be used directly.
+/// Precise semantics of data in this resource are not specified.
+#[derive(Resource, Default, Debug)]
+pub struct PrevState<S: States>(pub Option<S>);
+
 /// Run the enter schedule (if it exists) for the current state.
 pub fn run_enter_schedule<S: States>(world: &mut World) {
     world
         .try_run_schedule(OnEnter(world.resource::<State<S>>().0.clone()))
         .ok();
 }
 
-/// If a new state is queued in [`NextState<S>`], this system:
-/// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
-/// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
-/// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
-/// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
+/// Label to order all [`OnEnter`], [`OnTransition`], and [`OnExit`] systems.
+///
+/// All `OnEnter` will run before all `OnTransition`,
+/// and all `OnTransition` will run before all `OnExit`.
+///
+/// The order of `OnEnter` systems relative to each other (and so on) is ambiguous by default.
+#[derive(SystemSet, Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum StateTransitionSet {
+    /// Change all the states before invoking transition systems here.
+    BeforeTransition,
+    /// All [`OnEnter`] schedules are executed in this set.
+    OnEnter,
+    /// All [`OnTransition`] schedules are executed in this set.
+    OnTransition,
+    /// All [`OnExit`] schedules are executed in this set.
+    OnExit,
+    /// Cleanup `PrevState` after all transition systems have run.
+    AfterTransition,
+}
+
+/// Run all state transition systems for the given state type.
+///
+/// This function is deprecated. It is not used internally by Bevy,
+/// and will be removed in a future version.
+#[deprecated(since = "0.13.0")]
 pub fn apply_state_transition<S: States>(world: &mut World) {
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
+    apply_state_transition_before_transition::<S>(world);
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+    apply_state_transition_after_transition::<S>(world);
+}
+
+/// Change the state before invoking transition systems.
+pub fn apply_state_transition_before_transition<S: States>(world: &mut World) {
```

Comment:
This doesn't need to be an exclusive system anymore. We can get easy parallelism by accessing the state resources directly.

Reply:
To do this we need to kill legacy `apply_state_transition` first, which calls these systems (as functions) directly (or duplicate a lot of code).

This is much change than it seems.

Suggestions? Or we can do it in follow-up PR.

Reply:
I believe you could use `SystemState` to call this function from an exclusive system.

```rust
fn apply_state_transition_before_transition<S: States>(
    mut prev_state: ResMut<PrevState<S>>,
    mut state: ResMut<State<S>>,
    mut next_state: ResMut<NextState<S>>,
) {
...
}

fn apply_state_transition<S: States>(
    world: &mut World,
    before_transition_state: &mut SystemState<(
        ResMut<PrevState<S>>,
        ResMut<State<S>>,
        ResMut<NextState<S>>,
    )>,
) {
     let (prev, curr, next) = before_transition_state.get(world);
     apply_state_transition_before_transition(prev, curr, next);
}
```

Or if you don't want to repeat the function parameters you could use a type alias or custom `SystemParam`